### PR TITLE
Change reservoir sampling of segments to spliterator approach

### PR DESCRIPTION
**Pull request by shivtools**
**Issue**: https://github.com/apache/incubator-druid/issues/7147 

**Summary**: This PR attempts to make the sampling of a random segment more optimized by using a hybrid approach (suggested by @leventov) of reservoir sampling servers based on their populations along with the [following approach](https://stackoverflow.com/questions/12385284/how-to-select-a-random-key-from-a-hashmap-in-java/54893402#54893402) to pick a segment given a server.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 1]
```diff
@@ -21,36 +21,57 @@
...
-        // w.p. 1 / (numSoFar+1), swap out the server and segment
-        if (randNum == numSoFar) {
-          fromServerHolder = server;
-          proposalSegment = segment;
-        }
-        numSoFar++;
+
+    for (int i = 0; i < Math.log(segments.size()); i++) {
+      if (chosenSpliterator.estimateSize() < SPLITERATOR_SIZE_THRESHOLD) {
+        List<DataSegment> finalList = StreamSupport
```
Comment (drcrallen) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: can it just be checked if the spliterator is `SIZED` and pick a specific element if so?

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 2]
```diff
@@ -21,36 +21,57 @@
...
     if (fromServerHolder != null) {
       return new BalancerSegmentHolder(fromServerHolder.getServer(), proposalSegment);
     } else {
       return null;
     }
   }
 
+  static <T> T getRandomElementFromList(List<T> list)
+  {
+    if (list.size() == 0) {
```
Comment (drcrallen) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: Highly suggest `isEmpty`

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 3]
```diff
@@ -21,36 +21,57 @@
...
+      if (chosenSpliterator.estimateSize() < SPLITERATOR_SIZE_THRESHOLD) {
+        List<DataSegment> finalList = StreamSupport
+            .stream(chosenSpliterator, false)
+            .collect(Collectors.toList());
+
+        proposalSegment = getRandomElementFromList(finalList);
+        break;
+      }
+
+      Spliterator<DataSegment> newSpliterator = chosenSpliterator.trySplit();
```
Comment (drcrallen) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: > This method may return null for any reason, including emptiness, inability to split after traversal has commenced, data structure constraints, and efficiency considerations.

`null` needs handled

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 4]
```diff
@@ -21,36 +21,57 @@
...
-      for (DataSegment segment : server.getServer().getSegments()) {
-        int randNum = ThreadLocalRandom.current().nextInt(numSoFar + 1);
-        // w.p. 1 / (numSoFar+1), swap out the server and segment
-        if (randNum == numSoFar) {
-          fromServerHolder = server;
-          proposalSegment = segment;
-        }
-        numSoFar++;
+
+    for (int i = 0; i < Math.log(segments.size()); i++) {
```
Comment (drcrallen) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: what happens on an empty server?

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 5]
```diff
@@ -21,36 +21,57 @@
...
+import java.util.Collection;
 import java.util.List;
+import java.util.Spliterator;
 import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 final class ReservoirSegmentSampler
 {
+  final static int SPLITERATOR_SIZE_THRESHOLD = 10;
```
Comment (drcrallen) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: I suspect this can be a lot higher without penalty. Have you done any benchmarks?

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 4 above)
Comment (drcrallen) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: ((please add unit tests))

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 6]
```diff
@@ -21,36 +21,57 @@
...
+            .collect(Collectors.toList());
+
+        proposalSegment = getRandomElementFromList(finalList);
+        break;
+      }
+
+      Spliterator<DataSegment> newSpliterator = chosenSpliterator.trySplit();
+
+      // Choose between itself or the new spliterator with equal probability
+      if (ThreadLocalRandom.current().nextBoolean()) {
```
Comment (drcrallen) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: For testability I suggest passing in a `Random`, and making the method `@VisibleForTesting`, so that way you can have predictable results in unit tests. (Same with other places that inject randomness)

**Comment by jihoonson** [comment]
Hi @shivtools thank you for your contribution!

@leventov would you please take a look at this PR? I'm not sure this is what you want. Also, probably we need to add a new sampler rather than modifying the existing one because the changed algorithm looks not a reservoir sampling anymore.

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 5 above)
Comment (leventov) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: I would say that tests are needed to check this value ensures a good uniformity of choosing a segment in a HashMap. 10 might be too little. See [this comment](https://stackoverflow.com/questions/12385284/how-to-select-a-random-key-from-a-hashmap-in-java/54893402#comment96652203_54893402) on SO.

**Comment by leventov** [comment]
The algorithm can be optimized and still be reservoir sampling. See my comment in #7147.

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 4 above)
Comment (leventov) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: This is a natural logarithm, while you really want log2. `32 - Integer.numberOfLeadingZeros(size)` style of thing. Or `IntMath.log2()` from Guava.

**Comment by shivtools** [comment]
Thanks @leventov, @drcrallen and @jihoonson  for the review, will modify this soon and update the PR.

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 1 above)
Comment (shivtools) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: @drcrallen, thanks for the feedback! @leventov what are your thoughts about this? I think it would be easier than cleaner than `SPLITERATOR_SIZE_THRESHOLD` and having to benchmark to find a right number.

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 4 above)
Comment (shivtools) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: Do you mean a server which has no segments? I've added a test for this edge case.

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 5 above)
Comment (shivtools) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: I ran some benchmarks by hand and observed that `25` resulted in the most uniform selection of segments. See results in this [Gist](https://gist.github.com/shivtools/7e8af2038e155e2cd8131a10f53035a5).

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 1 above)
Comment (shivtools) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: Also, since my last comment, I've run benchmarks using different sizes for the spliterator threshold and observed a [pretty uniform selection of segments](https://gist.github.com/shivtools/7e8af2038e155e2cd8131a10f53035a5) using `25`.

**Comment by shivtools** [comment]
Hey @leventov, @drcrallen, I've made the changes you've requested. I'm seeing one Travis CI integration test fail with the following error, although I don't think it is related to my PR. Please let me know if otherwise: 

```
[ERROR] testRealtimeIndexTask(org.apache.druid.tests.indexer.ITRealtimeIndexTaskTest)  Time elapsed: 10.4 s  <<< FAILURE!
java.lang.RuntimeException: java.util.concurrent.ExecutionException: org.jboss.netty.channel.ChannelException: Faulty channel in resource pool
	at org.apache.druid.tests.indexer.ITRealtimeIndexTaskTest.postEvents(ITRealtimeIndexTaskTest.java:119)
	at org.apache.druid.tests.indexer.ITRealtimeIndexTaskTest.testRealtimeIndexTask(ITRealtimeIndexTaskTest.java:57)
Caused by: java.util.concurrent.ExecutionException: org.jboss.netty.channel.ChannelException: Faulty channel in resource pool
	at org.apache.druid.tests.indexer.ITRealtimeIndexTaskTest.postEvents(ITRealtimeIndexTaskTest.java:119)
	at org.apache.druid.tests.indexer.ITRealtimeIndexTaskTest.testRealtimeIndexTask(ITRealtimeIndexTaskTest.java:57)
Caused by: org.jboss.netty.channel.ChannelException: Faulty channel in resource pool
	at org.apache.druid.tests.indexer.ITRealtimeIndexTaskTest.postEvents(ITRealtimeIndexTaskTest.java:119)
	at org.apache.druid.tests.indexer.ITRealtimeIndexTaskTest.testRealtimeIndexTask(ITRealtimeIndexTaskTest.java:57)
Caused by: org.jboss.netty.channel.ChannelException: Failed to handshake with host[https://127.0.0.1:8101]
Caused by: java.nio.channels.ClosedChannelException
```

**Comment by jon-wei** [comment]
@shivtools That test is known to be flaky, I've restarted it

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 7]
```diff
@@ -19,36 +19,93 @@
...
+
+    if (segments.size() == 1) {
+      proposalSegment = segments.iterator().next();
+    } else {
+      Spliterator<DataSegment> chosenSpliterator = segments.spliterator();
+      if (chosenSpliterator == null) {
+        return null;
+      }
+
+      for (int i = 0; i < IntMath.log2(segments.size(), RoundingMode.UP); i++) {
```
Comment (drcrallen) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: can the logic here be flipped around like the following:

```java
while (chosenSpliterator.estimateSize() >= SPLITERATOR_SIZE_THRESHOLD) {
  Spliterator<DataSegment> newSpliterator = chosenSpliterator.trySplit();
  if (ThreadLocalRandom.current().nextBoolean()) {
    chosenSpliterator = newSpliterator;
  }
}
List<DataSegment> finalList = StreamSupport
              .stream(chosenSpliterator, false)
              .collect(Collectors.toList());
proposalSegment = getRandomElementFromList(finalList);
```

IMHO it is a bit easier to read

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 8]
```diff
@@ -19,36 +19,93 @@
...
+
+    ServerHolder sampledServer = null;
+    long total = 0;
+
+    for (ServerHolder serverHolder : serverHolders) {
+      ImmutableDruidServer server = serverHolder.getServer();
+      long numSegments = server.getSegments().size();
+      long endIndex = total + numSegments;
+      // Handle edge case where first server contains no segments so that nextLong() doesn't fail
+      long randomIndex = ThreadLocalRandom.current().nextLong((endIndex == 0) ? 1 : endIndex);
```
Comment (drcrallen) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: can you split this out explicitly in an if statement rather than having `nextLong(1)` as a nested consideration

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 5 above)
Comment (drcrallen) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: did you include stats tests here? there are random components at play so you can't claim most uniform without a measure for the randomness :)

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 7 above)
Comment (leventov) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: This is right. The loop doesn't need to be bounded at all. Also, a single `randomBits = nextInt()` can be computed before the loop and checking `((randomBits >>= 1) & 1) == 0` instead of calling into TLR every time.

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 5 above)
Comment (shivtools) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: Hey @drcrallen, I ran a very crude test that invokes `getRandomBalancerSegmentHolder()` 5000 times for different values of `spliterator size` and for each value of spliterator size, I calculated the variance (i.e `max - min`) across all segments about 200 times. I then used `.summaryStatistics()` to get an idea of what the variance looks like for each spliterator size. You can find my findings [here](https://gist.github.com/shivtools/f2766b6917dd381abd6330945a3baa49). 

In my opinion, as long as we _don't_ select `spliterator size = 2 or 3`, the variances obtained with other results look pretty similar. Is this what you had in mind?

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java (hunk 7 above)
Comment (shivtools) [reply server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: Thank you both for the helpful suggestions! Have incorporated both of these comments.

File: server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java [hunk 9]
```diff
@@ -19,36 +19,93 @@
...
-        if (randNum == numSoFar) {
-          fromServerHolder = server;
-          proposalSegment = segment;
+    ServerHolder fromServerHolder = reservoirSampleServer(serverHolders);
+    if (fromServerHolder == null) {
+      return null;
+    }
+
+    DataSegment proposalSegment;
+    Collection<DataSegment> segments = fromServerHolder.getServer().getSegments();
```
Comment (leventov) [inline server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java]: getSegments() returns a lazy collection (see `ImmutableDruidServer` code) that is not really suitable for spliteration. A random choice in `ImmutableDruidServer` is somewhat complicated by the fact that data is spread there across data sources, so you may need to do reservoir choice of a data source first, in the same way as you do reservoir choice of a server.

So there need to be two abstract, independently testable utilities:
 - `<T> T reservoirChooseSized(Collection<T> elements, ToIntFunction<T> getSize)`
 - `<K, V> Map.Entry<K, V> chooseRandomEntry(HashMap<K, V> m)`

But even if you do this, you cannot choose from an `ImmutableMap` in `ImmutableDruidDataSource` efficiently, because on the old Guava version that we use (16.0.1) `ImmutableMap` doesn't support efficient splitting.

It means that if you want to persist to bring this whole idea of optimized picking to live, you need to change the algorithm all the way from `DruidCoordinatorBalancer` and `BalancerStrategy.pickSegmentToMove()` and provide it with a list of `DruidServer` objects (borrowed from `DruidCoordinator` object) rather than `ImmutableDruidServer`, and eventually pick from `DruidDataSource.idToSegmentMap` (a `ConcurrentHashMap`).

This may be a daunting task. On the other hand, it may be drawn irrelevent by the refactoring of `DruidCoordinatorBalancer` that I plan for #7159. Therefore I suggest to put this effort on hold until after #7159 is fixed. Then it will be clearer if the problem is still relevant.
